plugins/obs-webrtc/webrtc-utils.h

```diff
@@ -0,0 +1,268 @@
+	std::vector<std::string> location_headers;
+
+	auto offer_sdp =
+		std::string(peer_connection->localDescription().value());
+
+	// Add user-agent to our requests
+	headers = curl_slist_append(headers, user_agent.c_str());
+
+	CURL *c = curl_easy_init();
+	curl_easy_setopt(c, CURLOPT_WRITEFUNCTION, curl_writefunction);
+	curl_easy_setopt(c, CURLOPT_WRITEDATA, (void *)&read_buffer);
+	curl_easy_setopt(c, CURLOPT_HEADERFUNCTION,
+			 curl_header_location_function);
+	curl_easy_setopt(c, CURLOPT_HEADERDATA, (void *)&location_headers);
+	curl_easy_setopt(c, CURLOPT_HTTPHEADER, headers);
+	curl_easy_setopt(c, CURLOPT_URL, endpoint_url.c_str());
+	curl_easy_setopt(c, CURLOPT_POST, 1L);
+	curl_easy_setopt(c, CURLOPT_COPYPOSTFIELDS, offer_sdp.c_str());
+	curl_easy_setopt(c, CURLOPT_TIMEOUT, 8L);
+	curl_easy_setopt(c, CURLOPT_FOLLOWLOCATION, 1L);
+	curl_easy_setopt(c, CURLOPT_UNRESTRICTED_AUTH, 1L);
+	curl_easy_setopt(c, CURLOPT_ERRORBUFFER, error_buffer);
+
+	auto cleanup = [&]() {
+		curl_easy_cleanup(c);
+		curl_slist_free_all(headers);
+	};
+
+	*curl_code = curl_easy_perform(c);
+	if (*curl_code != CURLE_OK) {
+		cleanup();
+		return webrtc_network_status::ConnectFailed;
+	}
+
+	long response_code;
+	curl_easy_getinfo(c, CURLINFO_RESPONSE_CODE, &response_code);
+	if (response_code != 201) {
+		cleanup();
+		return webrtc_network_status::InvalidHTTPStatusCode;
+	}
+
+	if (read_buffer.empty()) {
+		cleanup();
+		return webrtc_network_status::NoHTTPData;
+	}
+
+	long redirect_count = 0;
+	curl_easy_getinfo(c, CURLINFO_REDIRECT_COUNT, &redirect_count);
+
+	if (location_headers.size() < static_cast<size_t>(redirect_count) + 1) {
+		cleanup();
+		return webrtc_network_status::NoLocationHeader;
+	}
+
+	CURLU *url_builder = curl_url();
+	auto last_location_header = location_headers.back();
+
+	// If Location header doesn't start with `http` it is a relative URL.
+	// Construct a absolute URL using the host of the effective URL
+	if (last_location_header.find("http") != 0) {
+		char *effective_url = nullptr;
+		curl_easy_getinfo(c, CURLINFO_EFFECTIVE_URL, &effective_url);
+		if (effective_url == nullptr) {
+			cleanup();
+			return webrtc_network_status::FailedToBuildResourceURL;
+		}
+
+		curl_url_set(url_builder, CURLUPART_URL, effective_url, 0);
+		curl_url_set(url_builder, CURLUPART_PATH,
+			     last_location_header.c_str(), 0);
+		curl_url_set(url_builder, CURLUPART_QUERY, "", 0);
+	} else {
+		curl_url_set(url_builder, CURLUPART_URL,
+			     last_location_header.c_str(), 0);
+	}
+
+	char *url = nullptr;
+	CURLUcode rc = curl_url_get(url_builder, CURLUPART_URL, &url,
+				    CURLU_NO_DEFAULT_PORT);
+	if (rc) {
+		cleanup();
+		return webrtc_network_status::InvalidLocationHeader;
+	}
+
+	resource_url = url;
+	curl_free(url);
+	curl_url_cleanup(url_builder);
+
+	auto response = std::string(read_buffer);
+	response.erase(0, response.find("v=0"));
+
+	rtc::Description answer(response, "answer");
+	try {
+		peer_connection->setRemoteDescription(answer);
```

Comment:
Hey @Sean-Der, thank you for getting these changes going - we're excited to see more support of WHIP/WHEP in OBS. We're hoping to see full spec support (maybe too much to fit in this PR) where [`Link` headers are configured with the `peer_connection`](https://www.ietf.org/archive/id/draft-ietf-wish-whip-13.html#section-4.4) to handle STUN/TURN configs. Unfortunately as it currently stands, lack of STUN/TURN handling is a blocker for our customers being able to use OBS with WebRTC (either WHIP or WHEP).

If it can't fit into this PR, we have resources available to coordinate with you to make that change in a future PR.

Reply:
I am working on that right now @0xcadams, I am going to try to get it done this week :)

When you get a chance mind reviewing the [roadmap](https://github.com/obsproject/obs-studio/discussions/10372)? Would love to hear if I have any other gaps I am missing. 

I would be so grateful if I could get some dev support! Can you join the [Real-time Broadcasting Discord](https://discord.gg/An5jjhNUE3) that is where most coordination is taking place. I can split up/start the work right now!

Reply:
I do not think this PR should increase its scope further. Please keep its scope as-is unless you are fixing bugs.